compiler/passes/cleanup.cpp

```diff
@@ -67,19 +67,155 @@ void cleanup() {
 *                                                                             *
 ************************************** | *************************************/
 
+static bool areMultiDefExprsInAList(AList& list) { 
+   int numStmts = 0;
+
+    for_alist(stmt, list){
+      if (isDefExpr(stmt)) numStmts++;
+    }
+    
+    return numStmts > 1;
+}
+
+static void setAstHelp(Expr* parent, Expr*& lhs, Expr* rhs) {
+  if(lhs){
+    lhs->remove();
+  }
+  lhs = rhs;
+  parent_insert_help(parent, lhs);
+}
+
+static void backPropagateInFunction(BlockStmt* block) {
+
+  Expr* init = NULL;
+  DefExpr* prev = NULL;
+
+  if (!areMultiDefExprsInAList(block->body)) return;
+
+  SET_LINENO(block);
+
+  VarSymbol* typeTmp = NULL;
+  BlockStmt* tmpBlock = new BlockStmt();
+
+  for_alist_backward(stmt, block->body) {
+    if (DefExpr* def = toDefExpr(stmt)) {
+
+      //1. set local variableis -- analysis
+      if (def->init || def->exprType) {
+
+        if(def->init != NULL) {
+          init = def->init;
+        } else {
+          init = NULL;
+        }
+
+        typeTmp = NULL;
+
+      }
+
+      //2. update prev if necessary. Since the defExprs in a block statement are
+      //iterated in a reverse order, we use prev for cases like var x, y = 1.0, so
+      //that x can be initialized with 1.0 and y is initialized with the x symbol.
+      if (prev != NULL && def->init == NULL && def->exprType == NULL) {
+        SET_LINENO(prev);
+        if(prev->exprType != NULL && typeTmp == NULL){
+          typeTmp = newTemp("type_tmp");
+          typeTmp->addFlag(FLAG_TYPE_VARIABLE);
+          DefExpr* tmpDef = new DefExpr(typeTmp, prev->exprType->copy());
+          prev->exprType->replace(new SymExpr(typeTmp));
+          tmpBlock->insertAtTail(tmpDef);
+        }
+
+        if(prev->init != NULL && init != NULL){
+          if (init->isNoInitExpr()){
+            setAstHelp(prev, prev->init, init->copy());
+          } else if (typeTmp) {
+            setAstHelp(prev, prev->init, new CallExpr("chpl__readXX", new SymExpr(def->sym)));
+          } else {
+            setAstHelp(prev, prev->init, new SymExpr(def->sym));
+          }
+        }
+      }
+    
+      //3. update def, type then init
+      {
+        SET_LINENO(def);
+        if(typeTmp != NULL && def->exprType == NULL) {
+          setAstHelp(def, def->exprType, prev->exprType->copy());
+        }
+
+        if(init != NULL && def->init == NULL){
+          setAstHelp(def, def->init, init->copy());
+        }
+      }
+      prev = def;
+    }
+  }
+
+  if(tmpBlock) {
+    block->insertAtHead(tmpBlock);
+    tmpBlock->flattenAndRemove();
+  }	   
+}
+
+static void backPropagate(BaseAST* ast) {
+  if (BlockStmt* block = toBlockStmt(ast)) {
+    if (block->blockTag == BLOCK_SCOPELESS){
+      for_alist(stmt, block->body){
```

Comment:
It'd be nice to add a comment here indicating that we're looking for scopeless blocks containing only DefExprs (and end of statement markers) since these are how the multi-variable-declarations are represented after parsing